msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs

```diff
@@ -882,6 +882,40 @@ public void FindAccessToken_ScopeCaseInsensitive()
             Assert.IsTrue(item.ScopeSet.Contains(scopeInCache));
         }
 
+        [TestMethod]
+        [TestCategory("TokenCacheTests")]
+        public void CacheB2CTokenTest()
+        {
+            TokenCache B2CCache = new TokenCache();
+            var tenantID = "someTenantID";
+            var authority = Authority.CreateAuthority(String.Format("https://login.microsoftonline.com/tfp/{0}/somePolicy/oauth2/v2.0/authorize", tenantID), false);
+
+            MsalTokenResponse response = new MsalTokenResponse();
+            //creating IDToken with empty tenantID and displayableID/PreferedUserName for B2C scenario
+            response.IdToken = MockHelpers.CreateIdToken(String.Empty, String.Empty, String.Empty);
+            response.ClientInfo = MockHelpers.CreateClientInfo();
+            response.AccessToken = "access-token";
+            response.ExpiresIn = 3599;
+            response.CorrelationId = "correlation-id";
+            response.RefreshToken = "refresh-token";
+            response.Scope = TestConstants.Scope.AsSingleString();
+            response.TokenType = "Bearer";
+
+            RequestContext requestContext = new RequestContext(new MsalLogger(Guid.NewGuid(), null));
+            AuthenticationRequestParameters requestParams = new AuthenticationRequestParameters()
```

Comment:
What sets the authority type to B2C? Suggestion add an assert which validates that Authority Type is B2C (then you know the new code is run)

Reply:
@henrik-me The authority type is set internally by the Authority class. It checks the authority for different formats and determines the authority type from that. 
B2C authority's take the form of https://login.microsoftonline.com/tfp/someTenantID/somePolicy/oauth2/v2.0/authorize

Reply:
Great. then it should be no issue to validate that the B2C AuthorityType is set.

---
In reply to: [215021421](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215021421) [](ancestors = 215021421)

Reply:
Would still have linked the assert on the authority.type ... though of little value... also I suppose you latest changes would also work for missing PreferredUserName for non B2C flows. I suggest you add that test case as well

---
In reply to: [215088757](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215088757) [](ancestors = 215088757,215021421)